File: app/ios/src/StoreSearchOC/LandscapeViewController.m
//
//  LandscapeViewController.m
//  StoreSearchOC
//
//  Created by ZhangZehua on 2/23/17.
//  Copyright © 2017 ZhangZehua. All rights reserved.
//

#import "LandscapeViewController.h"
#import "DetailViewController.h"
#import "SearchResult.h"
#import "Search.h"

@interface LandscapeViewController () <UIScrollViewDelegate>

@property (nonatomic, weak) IBOutlet UIScrollView *scrollView;
@property (nonatomic, weak) IBOutlet UIPageControl *pageControl;

@end

@implementation LandscapeViewController {
  BOOL _firstTime;
}

- (id)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil {
  self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
  if (self) {
    _firstTime = YES;
  }
  return self;
}

- (void)viewWillLayoutSubviews {
  [super viewWillLayoutSubviews];
  if (_firstTime) {
    _firstTime = NO;
    if (self.search != nil) {
      if (self.search.isLoading) {
        [self showSpinner];
      } else if ([self.search.searchResults count] == 0) {
        [self showNothingFoundLabel];
      }else {
        [self tileButtons];
      }
    }
  }
}

- (void)viewDidLoad {
    [super viewDidLoad];
  self.scrollView.backgroundColor = [UIColor colorWithPatternImage:[UIImage imageNamed:@"LandscapeBackground"]];
  self.scrollView.contentSize = CGSizeMake(1000, self.scrollView.bounds.size.height);
  self.pageControl.numberOfPages = 0;
}

- (void)downloadImageForSearchResult:(SearchResult *)searchResult andPlaceOnButton:(UIButton *)button {
  NSURL *url = [NSURL URLWithString:searchResult.artworkURL60];
  [button setImage:[UIImage imageWithData:[NSData dataWithContentsOfURL:url]] forState:UIControlStateNormal];
}

- (void)showNothingFoundLabel {
  UILabel *label = [[UILabel alloc] initWithFrame:CGRectZero];
  label.text = NSLocalizedString(@"Nothing Found", @"Error alert: title");
  label.backgroundColor = [UIColor clearColor];
  label.textColor = [UIColor whiteColor];
  [label sizeToFit];
  CGRect rect = label.frame;
  rect.size.width = ceilf(rect.size.width / 2.0f) * 2.0f;
  rect.size.height = ceilf(rect.size.height / 2.0f) * 2.0f;
  label.frame = rect;
  label.center = CGPointMake(CGRectGetMidX(self.scrollView.bounds), CGRectGetMidY(self.scrollView.bounds));
  [self.view addSubview:label];
}

- (void)searchResultsReceived {
  [self hideSpinner];
  if ([self.search.searchResults count] == 0) {
    [self showNothingFoundLabel];
  } else {
    [self tileButtons];
  }
}

- (void)showSpinner {
  UIActivityIndicatorView *spinner = [[UIActivityIndicatorView alloc] initWithActivityIndicatorStyle:UIActivityIndicatorViewStyleWhiteLarge];
  spinner.center = CGPointMake(CGRectGetMidX(self.scrollView.bounds) + 0.5f, CGRectGetMidY(self.scrollView.bounds) + 0.5f);
  spinner.tag = 1000;
  [self.view addSubview:spinner];
  [spinner startAnimating];
}

- (void)hideSpinner
{
  [[self.view viewWithTag:1000] removeFromSuperview];
}

- (void)tileButtons
{
  int columnsPerPage = 5;
  CGFloat itemWidth = 96.0f;
  CGFloat x = 0.0f;
  CGFloat extraSpace = 0.0f;
  CGFloat scrollViewWidth = self.scrollView.bounds.size.width;
  if (scrollViewWidth > 480.0f) {
    columnsPerPage = 6;
    itemWidth = 94.0f;
    x = 2.0f;
    extraSpace = 4.0f;
  }
  const CGFloat itemHeight = 88.0f;
  const CGFloat buttonWidth = 82.0f;
  const CGFloat buttonHeight = 82.0f;
  const CGFloat marginHorz = (itemWidth - buttonWidth) / 2.0f;
  const CGFloat marginVert = (itemHeight - buttonHeight) / 2.0f;

  for (int index = 0, row = 0, column = 0; index < self.search.searchResults.count;) {
    UIButton *button = [UIButton buttonWithType:UIButtonTypeCustom];
    [self downloadImageForSearchResult:self.search.searchResults[index] andPlaceOnButton:button];
    [button setBackgroundImage:[UIImage imageNamed:@"LandscapeButton"] forState:UIControlStateNormal];
    button.frame = CGRectMake(x + marginHorz, 20.0f + row * itemHeight + marginVert, buttonWidth, buttonHeight);
    button.tag = 2000 + index;
    [button addTarget:self action:@selector(buttonPressed:)forControlEvents:UIControlEventTouchUpInside];
    [self.scrollView addSubview:button];
    index++;
    row++;
    if (row == 3) {
      row = 0;
      column++;
      x += itemWidth;
      if (column == columnsPerPage) {
        column = 0;
        x += extraSpace;
      }
    }
  }
  int tilesPerPage = columnsPerPage * 3;
  int numPages = ceilf([self.search.searchResults count] / (float)tilesPerPage);
  self.scrollView.contentSize = CGSizeMake(numPages * scrollViewWidth, self.scrollView.bounds.size.height);
  self.pageControl.numberOfPages = numPages;
  self.pageControl.currentPage = 0;
}

- (IBAction)pageChanged:(UIPageControl *)sender {
  [UIView animateWithDuration:0.3
                        delay:0
                      options:UIViewAnimationOptionCurveEaseInOut
                   animations:^{
                     self.scrollView.contentOffset = CGPointMake(self.scrollView.bounds.size.width * sender.currentPage, 0);
                   }
                   completion:nil];
}

- (void)buttonPressed:(UIButton *)sender {
  DetailViewController *controller = [[DetailViewController alloc] initWithNibName:@"DetailViewController" bundle:nil];
  SearchResult *searchResult = self.search.searchResults[sender.tag - 2000];
  controller.searchResult = searchResult;
  [controller presentInParentViewController:self];
}

- (void)scrollViewDidScroll:(UIScrollView *)theScrollView {
  CGFloat width = self.scrollView.bounds.size.width;
  int currentPage = (self.scrollView.contentOffset.x + width / 2.0f) / width;
  self.pageControl.currentPage = currentPage;
}

@end

File: app/ios/src/StoreSearchSwift/LandscapeViewController.swift
//
//  LandscapeViewController.swift
//  StoreSearch
//
//  Created by ZhangZehua on 3/3/16.
//  Copyright © 2016 ZhangZehua. All rights reserved.
//

import UIKit

class LandscapeViewController: UIViewController {

    @IBOutlet weak var pageControl: UIPageControl!
    @IBOutlet weak var scrollView: UIScrollView!
    
    var search : Search!
    private var firstTime = true
    private var downloadTasks = [NSURLSessionDownloadTask]()
    
    @IBAction func pageChanged(sender: UIPageControl) {
        UIView.animateWithDuration(0.3, delay: 0, options: .CurveEaseInOut, animations: {
            self.scrollView.contentOffset = CGPoint(
                x: self.scrollView.bounds.size.width * CGFloat(sender.currentPage),
                y: 0)
            }, completion: nil)
    }
    override func viewDidLoad() {
        super.viewDidLoad()

        view.removeConstraints(view.constraints)
        view.translatesAutoresizingMaskIntoConstraints = true
        
        pageControl.removeConstraints(pageControl.constraints)
        pageControl.translatesAutoresizingMaskIntoConstraints = true
        pageControl.numberOfPages = 0
        
        scrollView.removeConstraints(scrollView.constraints)
        scrollView.translatesAutoresizingMaskIntoConstraints = true
        
        scrollView.backgroundColor = UIColor(patternImage: UIImage(named: "LandscapeBackground")!)
    }

    override func viewWillLayoutSubviews() {
        super.viewWillLayoutSubviews()
        
        scrollView.frame = view.bounds
        
        pageControl.frame = CGRect(x: 0,
            y: view.frame.size.height - pageControl.frame.size.height,
            width: view.frame.size.width,
            height: pageControl.frame.size.height)
        
        if firstTime {
            firstTime = false
            
            switch search.state {
            case .NotSearchedYet:
                break
            case .Loading:
                showSpinner()
            case .NoResults:
                showNothingFoundLabel()
            case .Results(let list):
                tileButtons(list)
            }
        }
    }
    
    func searchResultsReceived() {
        hideSpinner()
        
        switch search.state {
        case .NotSearchedYet, .Loading:
            break
        case .NoResults:
            showNothingFoundLabel()
        case .Results(let list):
            tileButtons(list)
        }
    }
    
    private func hideSpinner() {
        view.viewWithTag(1000)?.removeFromSuperview()
    }
    
    private func showSpinner() {
        let spinner = UIActivityIndicatorView(activityIndicatorStyle: .WhiteLarge)
        spinner.center = CGPoint(
            x: CGRectGetMidX(scrollView.bounds) + 0.5,
            y: CGRectGetMidY(scrollView.bounds) + 0.5)
        spinner.tag = 1000
        view.addSubview(spinner)
        spinner.startAnimating()
    }
    
    private func showNothingFoundLabel() {
        let label = UILabel(frame: CGRect.zero)
        label.text = NSLocalizedString("Nothing Found", comment: "No items found of current searched content")
        label.backgroundColor = UIColor.clearColor()
        label.textColor = UIColor.whiteColor()
        
        label.sizeToFit()
        
        var rect = label.frame
        rect.size.width = ceil(rect.size.width / 2) * 2 // make even
        rect.size.height = ceil(rect.size.height / 2) * 2 // make even
        label.frame = rect
        
        label.center = CGPoint(
            x: CGRectGetMidX(scrollView.bounds),
            y: CGRectGetMidY(scrollView.bounds))
        view.addSubview(label)
    }
    
    private func tileButtons(searchResults: [SearchResult]) {
        // scrollViewWidth 480 iPhone 4S
        var columnsPerPage = 5
        var rowsPerPage = 3
        
        var itemWidth: CGFloat = 96
        var itemHeight: CGFloat = 88
        var marginX: CGFloat = 0
        var marginY: CGFloat = 20
        
        let scrollViewWidth = scrollView.bounds.size.width
        
        switch scrollViewWidth {
        case 568:   // iPhone 5
            columnsPerPage = 6
            itemWidth = 94
            marginX = 2
        case 667:   // iPhone 6
            columnsPerPage = 7
            itemWidth = 95
            itemHeight = 98
            marginX = 1
            marginY = 29
        case 736:   // iPhone 6 Plus
            columnsPerPage = 8
            rowsPerPage = 4
            itemWidth = 92
        default:
            break
        }
        
        let buttonWidth: CGFloat = 82
        let buttonHeight: CGFloat = 82
        let paddingHorz = (itemWidth - buttonWidth) / 2
        let paddingVert = (itemHeight - buttonHeight) / 2
        
        var row = 0
        var column = 0
        var x = marginX
        
        for (index, searchResult) in searchResults.enumerate() {
            let button = UIButton(type: .Custom) as UIButton
            button.setBackgroundImage(UIImage(named: "LandscapeButton"), forState: .Normal)
            downloadImageForSearchResult(searchResult, andPlaceOnButton: button)
            
            button.frame = CGRect(
                x: x + paddingHorz,
                y: marginY + CGFloat(row) * itemHeight + paddingVert,
                width: buttonWidth,
                height: buttonHeight)
            
            button.tag = 2000 + index
            button.addTarget(self, action: Selector("buttonPressed:"), forControlEvents: .TouchUpInside)
            
            scrollView.addSubview(button)
            ++row
            if row == rowsPerPage {
                row = 0
                ++column
                x += itemWidth
                
                if column == columnsPerPage {
                    column = 0
                    x += marginX * 2
                }
            }
        }
        
        let buttonsPerPage = columnsPerPage * rowsPerPage
        let numPages = 1 + (searchResults.count - 1) / buttonsPerPage
        
        scrollView.contentSize = CGSize(
            width: CGFloat(numPages) * scrollViewWidth,
            height: scrollView.bounds.size.height)
    
        print("Number of pages: \(numPages)")
        
        pageControl.numberOfPages = numPages
        pageControl.currentPage = 0
    }
    
    func buttonPressed(sender: UIButton) {
        performSegueWithIdentifier("ShowDetail", sender: sender)
    }
    
    override func prepareForSegue(segue: UIStoryboardSegue, sender: AnyObject?) {
        if segue.identifier == "ShowDetail" {
            switch search.state {
            case .Results(let list):
                let detailViewController = segue.destinationViewController as! DetailViewController
                let searchResult = list[sender!.tag - 2000]
                detailViewController.searchResult = searchResult
                
                detailViewController.isPopUp = true
            default:
                break
            }
        }
    }
    
    private func downloadImageForSearchResult(searchResult: SearchResult, andPlaceOnButton button: UIButton) {
        if let url = NSURL(string: searchResult.artworkURL60) {
            let session = NSURLSession.sharedSession()
            let downloadTask = session.downloadTaskWithURL(url, completionHandler: {
                [weak button] url, response, error in
                if error == nil && url != nil {
                    if let data = NSData(contentsOfURL: url!) {
                        if let image = UIImage(data: data) {
                            dispatch_async(dispatch_get_main_queue()) {
                                if let button = button {
                                    button.setImage(image, forState: .Normal)
                                }
                            }
                        }
                    }
                }
            })
            downloadTask.resume()
            downloadTasks.append(downloadTask)
        }
    }
    
    deinit {
        print("deinit \(self)")
        
        for task in downloadTasks {
            task.cancel()
        }
    }

    /*
    // MARK: - Navigation

    // In a storyboard-based application, you will often want to do a little preparation before navigation
    override func prepareForSegue(segue: UIStoryboardSegue, sender: AnyObject?) {
        // Get the new view controller using segue.destinationViewController.
        // Pass the selected object to the new view controller.
    }
    */

}

extension LandscapeViewController: UIScrollViewDelegate {
    func scrollViewDidScroll(scrollView: UIScrollView) {
        let width = scrollView.bounds.size.width
        let currentPage = Int((scrollView.contentOffset.x + width / 2) / width)
        pageControl.currentPage = currentPage
    }
}
